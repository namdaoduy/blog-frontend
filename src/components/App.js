import React, { Component } from 'react';
import { BrowserRouter as Router, Switch, Route, Redirect } from 'react-router-dom'
import Loadable from 'react-loadable';
import CssBaseline from '@material-ui/core/CssBaseline';
import JssProvider from 'react-jss/lib/JssProvider';
import { create } from 'jss';
import { createGenerateClassName, jssPreset } from '@material-ui/core/styles';

import history from 'utils/history';

const generateClassName = createGenerateClassName();
const jss = create({
  ...jssPreset(),
  insertionPoint: 'jss-insertion-point',
});

const LoadingComponent = name => ({ isLoading, error }) => {
  if (isLoading) {
    console.log(`Loading component... ${name}`);
  } else if (error) {
    console.log(`Loading component error ${name}`, error);
  }
  return null;
};

const Home = Loadable({
  loader: () => import('./Home'),
  loading: LoadingComponent('Home'),
});

const Blog = Loadable({
  loader: () => import('./Blog'),
  loading: LoadingComponent('Blog'),
});

const Login = Loadable({
  loader: () => import('./Login'),
  loading: LoadingComponent('Login'),
});

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLogin: false
    }
  }

  renderRoutes() {
    const routes = [
      <Route path="/login" component={Login} key="login"/>,
      <Route path="/blog/:blog_id" component={Blog} key="blog"/>,
      <Route path="/" component={Home} key="home"/>,
      <Redirect to="/" key="redirect"/>
    ]

    if (!this.state.isLogin) {
      return routes;
    }
    else {
      return [...routes,
        <Route path="/user/blogs" component={Home} key="blogs"/>
      ]
    }
  }

  render() {
    return (
      <JssProvider jss={jss} generateClassName={generateClassName}>
        <React.Fragment>
          <CssBaseline />
          <Router>
            <Switch>
              {this.renderRoutes()}
            </Switch>
          </Router>
        </React.Fragment>
      </JssProvider>
    );
  }
}