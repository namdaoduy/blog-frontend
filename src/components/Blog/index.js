import Avatar from '@material-ui/core/Avatar';
import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import Toolbar from '@material-ui/core/Toolbar';
import Tooltip from '@material-ui/core/Tooltip';
import Typography from '@material-ui/core/Typography';
import Favorite from '@material-ui/icons/Favorite';
import FavoriteBorder from '@material-ui/icons/FavoriteBorder';
import Share from '@material-ui/icons/Share';
import React, { Component } from 'react';
import Markdown from 'react-markdown';
import Moment from 'react-moment';
import { connect } from 'react-redux';
import { getBlogById, likeBlog } from '../../actions/blog';
import '../../assets/styles/blog.css';
import Header from '../Common/Header';

class Blog extends Component {
  constructor(props) {
    super(props);
    const { match } = this.props;
    this.state = {
      id: match.params.blog_id,
    };
  }

  componentDidMount() {
    this.fetchBlog();
  }

  fetchBlog = () => {
    const { id } = this.state;
    this.props.getBlogById(id)
      .then(res => console.log(res.success))
      .catch(err => console.log(err));
  };

  fetchLike = () => {
    const { id } = this.state;
    this.props.likeBlog(id)
      .then(res => console.log(res.success))
      .catch(err => console.log(err));
  }

  onLike = () => {
    const { loggedIn, isLiked } = this.props;
    if (!loggedIn) {
      alert('You have to log in to like blogs.');
    }
    if (isLiked) {
      return;
    }
    this.fetchLike();
  }

  render() {
    const { blogData, isLiked } = this.props;
    return (
      <div className="blog-container">
        <Header />
        <Grid className="blog-inner" container spacing={24}>
          <Grid item xs={12}>
            <Typography variant="h3" className="serif-2">
              {blogData.title || 'Blog title'}
            </Typography>

            <div className="blog-info">
              <Avatar
                className="author-avatar"
                alt="Author Avatar"
                src={blogData.picture}
              />
              <div className="blog-info-text">
                <Typography variant="subtitle2">
                  {blogData.author || "Author's Name"}
                </Typography>
                <Typography variant="caption">
                  <Moment fromNow date={blogData.created_at || 1545813100264} />
                  <span className="dot-divider" />
                  {`${blogData.timeRead || 10} min read`}
                </Typography>
              </div>
            </div>

            <div className="blog-content">
              <div className="blog-content-left">
                <Toolbar className="blog-btn-list">
                  <Tooltip title={blogData.like || 0} placement="left">
                    <IconButton
                      color="secondary"
                      onClick={this.onLike}
                    >
                      {isLiked ? <Favorite /> : <FavoriteBorder />}
                    </IconButton>
                  </Tooltip>
                  <IconButton color="secondary">
                    <Share />
                  </IconButton>
                </Toolbar>
              </div>
              <div className="blog-content-right">
                <Markdown source={blogData.body} className="blog-markdown" />
              </div>
            </div>
          </Grid>
        </Grid>
      </div>
    );
  }
}

const mapStateToProps = ({ blog, user }) => ({
  blogData: blog.blogData,
  isLiked: blog.isLiked,
  loggedIn: user.loggedIn,
});

const mapDispatchToProps = {
  getBlogById,
  likeBlog,
};

export default connect(mapStateToProps, mapDispatchToProps)(Blog);
