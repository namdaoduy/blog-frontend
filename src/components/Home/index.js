import Button from '@material-ui/core/Button';
import Divider from '@material-ui/core/Divider';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import React, { Component } from 'react';
import '../../assets/styles/home.css';
import API from '../../services/apis';
import history from '../../utils/history';
import Header from '../Common/Header';
import NewBlog from './NewBlog';
import TrendingBlog from './TrendingBlog';

export default class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      newBlogs: [],
      trendingBlogs: [],
    };
  }

  componentDidMount() {
    this.fetchNewBlogs();
    this.fetchTrendingBlogs();
  }

  handleWriteNow = () => {
    history.push('/user/blog/new');
  }

  handleMakeAccount = () => {
    history.push('/login');
  }

  renderNewBlogs = () => {
    const { newBlogs } = this.state;
    return newBlogs.map(blog => (
      <NewBlog blog={blog} key={blog.id} />
    ));
  }

  renderTrendingBlogs = () => {
    const { trendingBlogs } = this.state;
    return trendingBlogs.map((blog, i) => (
      <TrendingBlog blog={blog} key={blog.id} rank={i + 1} />
    ));
  }

  fetchNewBlogs = () => {
    API.getAllBlogs()
      .then((res) => {
        if (!res.success) return;
        this.setState({ newBlogs: res.data });
      })
      .catch(err => console.log(err));
  }

  fetchTrendingBlogs = () => {
    API.getTrendingBlogs()
      .then((res) => {
        if (!res.success) return;
        this.setState({ trendingBlogs: res.data });
      })
      .catch(err => console.log(err));
  }

  render() {
    return (
      <div className="home-container">
        <Header />
        <Grid
          className="home-inner"
          container
          spacing={24}
        >
          <Grid item xs={12}>
            <Paper
              className="home-welcome"
              elevation={5}
            >
              <Typography className="serif" variant="h1" gutterBottom>
                Welcome to Just Blog,
                {' '}
                <br />
                write your story.
              </Typography>
              <Typography variant="subtitle1" gutterBottom>
                Weâ€™ll deliver the best stories and ideas on the topics you care about most straight to your homepage, app, or inbox.
              </Typography>
              <Button
                className="btn-1"
                variant="contained"
                color="secondary"
                onClick={this.handleWriteNow}
              >
                Write Now
              </Button>
              <Button
                className="btn-2"
                variant="outlined"
                color="secondary"
                onClick={this.handleMakeAccount}
              >
                Make account in seconds
              </Button>
            </Paper>
          </Grid>

          <Grid item xs={8}>
            <Typography variant="h5" className="semi-bold serif-2">
              New blogs
            </Typography>
            <Divider />
            <Grid container spacing={24}>
              {this.renderNewBlogs()}
            </Grid>
          </Grid>

          <Grid item xs={4}>
            <Typography variant="h5" className="semi-bold serif-2">
              Trending
            </Typography>
            <Divider />
            <Grid container spacing={16}>
              {this.renderTrendingBlogs()}
            </Grid>
          </Grid>

        </Grid>
      </div>
    );
  }
}