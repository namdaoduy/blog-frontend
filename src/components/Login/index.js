import React, { Component } from 'react';
import '../../assets/styles/login.css';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import { GoogleLogin } from 'react-google-login';
import configs from '../../configs';
import API from '../../services/apis';

export default class Login extends Component {
  responseGoogle = (res) => {
    console.log(res);
    if (res.error) {
      return alert('Login Failed');
    }

    API.postLogin(res)
      .then((res) => {
        this.props.handleLogin(res);
      })
      .catch(err => console.log(err));
  }


  render() {
    return (
      <div className="login-container">
        <Typography variant="h2" className="serif" gutterBottom>
          Just Blog
        </Typography>
        <Paper className="login-paper" elevation={8}>
          <Typography variant="h4" className="serif-2" gutterBottom>
            Join us.
          </Typography>
          <Typography variant="subtitle1" gutterBottom>
            Create an account to personalize your homepage, follow your favorite authors and publications, applaud stories you love, and more.
          </Typography>
          <GoogleLogin
            clientId={configs.googleClientConfigs.web.client_id}
            scope="profile email"
            buttonText="Login with Google"
            onSuccess={this.responseGoogle}
            onFailure={this.responseGoogle}
          />
        </Paper>
      </div>
    );
  }
}