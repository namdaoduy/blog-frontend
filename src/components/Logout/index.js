import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';

export default class Logout extends Component {
  logout = () => {
    localStorage.removeItem('justblog_access_token');
    localStorage.removeItem('justblog_user_id');
    localStorage.removeItem('justblog_login_state');
    return null;
  }

  render() {
    return (
      <div>
        {this.logout()}
        <Redirect to="/" />
      </div>
    );
  }
}