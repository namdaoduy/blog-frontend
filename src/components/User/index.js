import React, { Component } from 'react';
import '../../assets/styles/user.css';
import { Switch, Route } from 'react-router-dom';
import { Grid } from '@material-ui/core';
import { MuiThemeProvider } from '@material-ui/core/styles';
import Header from '../Common/Header';
import theme from '../../constants/theme';
import Profile from './Profile';
import BlogEditor from './BlogEditor';
import API from '../../services/apis';

export default class User extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user_id: localStorage.getItem('justblog_user_id'),
      blogs: [],
    };
  }

  componentDidMount() {
    this.fetchUserBlogs();
  }

  fetchUserBlogs = () => {
    API.getBlogsByUser(this.state.user_id)
      .then((res) => {
        this.setState({ blogs: res.data });
      });
  }

  handleDelete = (id) => {
    const del = window.confirm('Are you sure to delete?');
    if (!del) return;
    const token = localStorage.getItem('justblog_access_token');
    API.deleteBlogById(id, token)
      .then((res) => {
        window.location.reload();
      })
      .catch(err => console.log(err));
  }

  handleClickEdit = (id) => {
    this.props.history.push(`/user/blog/${id}`);
  }

  handleClickNew = () => {
    this.props.history.push('/user/blog/new');
  }

  handleNew = (data) => {
    const token = localStorage.getItem('justblog_access_token');
    API.postBlog(data.title, data.body, token)
      .then((res) => {
        console.log(res);
        this.props.history.push('/user');
      });
  }

  handleEdit = (data) => {
    const token = localStorage.getItem('justblog_access_token');
    API.putBlog(data.id, data.title, data.body, token)
      .then((res) => {
        console.log(res);
        this.props.history.push('/user');
      });
  }

  render() {
    return (
      <MuiThemeProvider theme={theme}>
        <div className="user-container">
          <Header />
          <Grid
            className="user-inner"
            container
            spacing={24}
          >
            <Switch>
              <Route
                path="/user/blog/new"
                render={props => (
                  <BlogEditor
                    {...props}
                    type="new"
                    key={`blogNew${Date.now()}`}
                    handleNew={this.handleNew.bind(this)}
                  />
                )}
              />
              <Route
                path="/user/blog/:blog_id"
                render={props => (
                  <BlogEditor
                    {...props}
                    type="edit"
                    key={`blogEdit${Date.now()}`}
                    handleEdit={this.handleEdit.bind(this)}
                  />
                )}
              />
              <Route
                path="/user"
                render={props => (
                  <Profile
                    {...props}
                    key={`profile${Date.now()}`}
                    handleClickNew={this.handleClickNew.bind(this)}
                    handleClickEdit={this.handleClickEdit.bind(this)}
                    handleDelete={this.handleDelete}
                    handleEdit={this.handleEdit}
                    blogs={this.state.blogs}
                  />
                )}
              />
            </Switch>
          </Grid>
        </div>
      </MuiThemeProvider>
    );
  }
}