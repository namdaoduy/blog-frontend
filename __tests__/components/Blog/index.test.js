import React from 'react';
import { shallow } from 'enzyme';
import { Blog } from 'components/Blog';
import Favorite from '@material-ui/icons/Favorite';
import FavoriteBorder from '@material-ui/icons/FavoriteBorder';

describe('components/Blog/index', () => {
  let wrapper;
  let props;

  const setup = () => {
    wrapper = shallow(<Blog {...props} />);
  };

  beforeEach(() => {
    props = {
      match: {
        params: {
          blog_id: 1,
        },
      },
      blogData: {},
      isLiked: false,
      loggedIn: false,
      getBlogById: jest.fn(async () => {}),
      likeBlog: jest.fn(async () => {}),
    };
  });

  it('Render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('Test like button', () => {
    it('When not logged in, expect alert error', () => {
      props.loggedIn = false;
      setup();
      const btnLike = wrapper.find('.btn-like');
      btnLike.simulate('click');
      expect(window.alert).toHaveBeenCalledTimes(1);
    });

    it('When logged in, but liked, return', () => {
      props.loggedIn = true;
      props.isLiked = true;
      setup();
      const instance = wrapper.instance();
      const mockLike = jest.spyOn(instance, 'fetchLike');
      const btnLike = wrapper.find('.btn-like');
      instance.componentDidMount();
      btnLike.simulate('click');
      expect(mockLike).toHaveBeenCalledTimes(0);
    });

    it('When logged in, not liked before, call fetchLike', () => {
      props.loggedIn = true;
      props.isLiked = false;
      setup();
      const instance = wrapper.instance();
      const mockLike = jest.spyOn(instance, 'fetchLike');
      const btnLike = wrapper.find('.btn-like');
      instance.componentDidMount();
      btnLike.simulate('click');
      expect(mockLike).toHaveBeenCalledTimes(1);
    });
  });

  describe('Test render like button', () => {
    it('When liked', () => {
      props.isLiked = true;
      setup();
      expect(wrapper.find(Favorite).exists()).toBe(true);
    });

    it('When not liked', () => {
      props.isLiked = false;
      setup();
      expect(wrapper.find(FavoriteBorder).exists()).toBe(true);
    });
  });
});