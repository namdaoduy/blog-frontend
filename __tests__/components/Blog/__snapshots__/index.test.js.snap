// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Blog/index Render correctly 1`] = `
<div
  className="blog-container"
>
  <Connect(Header) />
  <WithStyles(Grid)
    className="blog-inner"
    container={true}
    spacing={24}
  >
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <WithStyles(Typography)
        className="serif-2"
        variant="h3"
      >
        Blog title
      </WithStyles(Typography)>
      <div
        className="blog-info"
      >
        <WithStyles(Avatar)
          alt="Author Avatar"
          className="author-avatar"
        />
        <div
          className="blog-info-text"
        >
          <WithStyles(Typography)
            variant="subtitle2"
          >
            Author's Name
          </WithStyles(Typography)>
          <WithStyles(Typography)
            variant="caption"
          >
            <t
              ago={false}
              calendar={false}
              date={1545813100264}
              decimal={false}
              element="time"
              filter={[Function]}
              fromNow={true}
              interval={60000}
              onChange={[Function]}
              titleFormat=""
              toNow={false}
              unit={null}
              unix={false}
              utc={false}
              withTitle={false}
            />
            <span
              className="dot-divider"
            />
            10 min read
          </WithStyles(Typography)>
        </div>
      </div>
      <div
        className="blog-content"
      >
        <div
          className="blog-content-left"
        >
          <WithStyles(Toolbar)
            className="blog-btn-list"
          >
            <WithStyles(Tooltip)
              placement="left"
              title={0}
            >
              <WithStyles(IconButton)
                className="btn-like"
                color="secondary"
                onClick={[Function]}
              >
                <pure(FavoriteBorderIcon) />
              </WithStyles(IconButton)>
            </WithStyles(Tooltip)>
            <WithStyles(IconButton)
              color="secondary"
            >
              <pure(ShareIcon) />
            </WithStyles(IconButton)>
          </WithStyles(Toolbar)>
        </div>
        <div
          className="blog-content-right"
        >
          <ReactMarkdown
            astPlugins={Array []}
            className="blog-markdown"
            escapeHtml={true}
            plugins={Array []}
            rawSourcePos={false}
            renderers={Object {}}
            skipHtml={false}
            sourcePos={false}
            transformLinkUri={[Function]}
          />
        </div>
      </div>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</div>
`;
